Add `IResourceBuilder<T>` extension method for event subscriptions

afscrome opened the pull request:
## Description

Add a number of extension method on `IResourceBuilder<T>` to easily subscribe to `IDistributedApplicationResourceEvent` on a resource

Refactored a few subscription calls over to use the new methods.  I can refactor more, just wanted to get a small amount (one per event) to prove the concept out.

Fixes #9168

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [x] No. Follow-up changes expected. - Follow up PR to migrate existing uses of `.Subscribe<T>` over to the new methods
- Are you including unit tests for the changes and scenario tests if relevant?
  - [ ] Yes
  - [X] No
- Did you add public API?
  - [X] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [X] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [X] Yes
      - [ ] No
  - [ ] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [X] No
- Does the change require an update in our Aspire docs?
  - [X] Yes
    - Link to aspire-docs issue (consider using one of the following templates):
      - [New (or update) `doc-idea` template](https://github.com/dotnet/docs-aspire/issues/new?template=02-docs-request.yml)
      - [New `breaking-change` template](https://github.com/dotnet/docs-aspire/issues/new?template=04-breaking-change.yml)
      - [New `diagnostic` template](https://github.com/dotnet/docs-aspire/issues/new?template=06-diagnostic-addition.yml)
  - [ ] No

afscrome committed:
Add IResourceBuilder<T> extension method for event subscriptions

Add a number of extension method on `IResourceBuilder<T>` to easily subscribe to `IDistributedApplicationResourceEvent` on a resource

Refactored a few subscription calls over to use the new methods - at least one for each event

afscrome commented:
cc @DamianEdwards

afscrome committed:
Correct `OnResourceReady` type restriction

afscrome committed:
Migrate a few more cases to the new API.

afscrome committed:
Add `resource` argument to resource subscription extension method callback.

mitchdenny commented:
Yeah, I think `On*` already is an established pattern in .NET and I'd prefer not to muddy the waters there anymore (there is already some inconsistency there that we could make worse by doing this with Aspire).

I think what we want is `When<T>(...)` and `WhenEvent(...)` which calls `When<T>(...)`.

davidfowl reviewed (approved):
Love it! We can do a follow up to finishing consuming these APIs everywhere

afscrome committed:
PR Feedback

- Fix formatting in AzureCosmosDBExtensions
- Fix rogue `"/>` in `On*` method xml docs
